File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int result = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        /*
        if null return 0;
        find left sum
        find right sum
        add val + left + right sum.
        update result;
        */
        treeSum(root);
        return result;
    }

    public int treeSum(TreeNode root) {
        if(root == null) return 0;
        int leftSum = treeSum(root.left);
        int rightSum = treeSum(root.right);
        leftSum = leftSum < 0 ? 0 : leftSum;
        rightSum = rightSum < 0 ? 0 : rightSum;
        result = Math.max(result, leftSum + rightSum + root.val); 
        return Math.max(leftSum, rightSum) + root.val;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        return self.helper(root, root.val)

    def helper(self, root: Optional[TreeNode], maxVal: int) -> int:
        if not root:
            return 0
        left = self.helper(root.left, max(maxVal, root.val))
        right = self.helper(root.right, max(maxVal, root.val))
        if(root.val >= maxVal):
            return left + right + 1
        else:
            return left + right